Move [role] to lua (fix for bug #23630)

Elvish-Hunter opened the pull request:
This PR converts the [role] tag from C++ to Lua.
In order to ensure that the patch has the intended effect, I also added a new unit test which checks if the role is assigned according to the order specified in the type= list.

gfgtdf commented:
Would this still work if i pass a config instead of a vconfig? Like 
```
wesnoth.wml_actions.role { role = "role1", type = "Spearman" }
```

Elvish-Hunter commented:
> Would this still work if i pass a config instead of a vconfig?

Good catch. It didn't work from Lua, so I fixed it. The issue was caused by the fact that I used `cfg.__shallow_literal` instead of `helper.shallow_literal`.
After that, I found out that calling the tag from Lua caused the `type=` and `role=` keys to be deleted too early, so I fixed that as well.
I also added a new unit test, to be sure that `wesnoth.wml_actions.role` works correctly when called from Lua.

gfgtdf commented:
looks good to me now
